### ISSUE TITLE
`have_enqueued_job#with` does not evaluate a `do ... end` block

### ISSUE THREAD

**Issue by benwalsh** [rspec/rspec-rails#1780]
When passing a block to `with`, it is only evaluated if it is in the higher-priority `{ ... }` syntax, and not if it is in `do ... end`

```
expect do
  create(:thing)
end.to have_enqueued_job(ThingCreationJob).with { |message|
  # This works and the expectation is tested
  expect(message['body']).to eq('A thing has been created')
}
```

```
expect do
  create(:thing)
end.to have_enqueued_job(ThingCreationJob).with do |message|
  # Code in this block is not evaluated, this expectation is not tested
  expect(message['body']).to eq('A thing has been created')
end
```

RSpec 3.5.4, Rails 4.2.7.1

**Comment by JonRowe** [comment]
Yep sorry, this is a Ruby thing, we can't control where blocks are sent, so we recommend only ever using `{ ... }` blocks with `expect`

**Comment by samphippen** [comment]
@JonRowe to which call does the second do block get routed? Could we pull some shenanigans here?

**Comment by NobodysNightmare** [comment]
This issue helped me figure out why my specs did not work as I intended them to, and I learned something new about Ruby operator precedence. Thank you! :)

What I am wondering now is why this code **is** working as intended:

```ruby
allow(MyJob).to receive(:perform_later) do ... end
```

From what I've learned today, the `do ... end` block is not sent to `receive`, but it will be sent to `to`. Still the code works as intended.

Are there any "tricks" implemented between `to` and `receive` to make this possible or am I overlooking another piece of language nuance?
